### PR TITLE
feat(accessibility): add tabbing to chart menu in dashboard

### PR DESCRIPTION
**Pull request by eschutho**
### SUMMARY

This change adds the ability to tab through elements on a dashboard, the title, charts, and into the chart controls, and execute each of them with a keyboard. 

https://github.com/apache/superset/assets/5186919/4dd70521-131a-438d-b370-ab7557e1176b

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by yousoph** [comment]
When using keyboard navigation to tab through on a dashboard, the highlight on the chart title in dashboards doesn't seem to be appearing for me

**Comment by kasiazjc** [comment]
> @eschutho Ephemeral environment spinning up at http://35.87.72.46:8080. Credentials are `admin`/`admin`. Please allow several minutes for bootstrapping and startup.

I think this looks good! Thanks @eschutho 🙏

File: superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts [hunk 1]
```diff
@@ -249,9 +249,13 @@ describe('Drill to detail modal', () => {
       it('opens the modal with the correct filters', () => {
         interceptSamples();
 
+        // focus on table first to trigger browser scroll
+        cy.get("[data-test-viz-type='table']").contains('boy').rightclick();
+
+        cy.wait(500);
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts:255]: I wonder if there's something we can do with timeouts instead of waiting, but... probably not worth the hassle.

File: superset-frontend/src/components/PageHeaderWithActions/index.tsx [hunk 2]
```diff
@@ -43,6 +43,9 @@ export const menuTriggerStyles = (theme: SupersetTheme) => css`
   &:hover:not(:focus) > span.anticon {
     color: ${theme.colors.primary.light1};
   }
+  &:focus-visible {
+    outline: 2px solid ${theme.colors.primary.dark2};
```
Comment (rusackas) [inline superset-frontend/src/components/PageHeaderWithActions/index.tsx:47]: One of these years I'm going to add some borderWidth properties to the theme...

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx [hunk 3]
```diff
@@ -233,6 +234,18 @@ const DashboardContainer: FC<DashboardContainerProps> = ({ topLevelTabs }) => {
...
             allowOverflow
+            onFocus={e => {
+              if (
+                // prevent scrolling when tabbing to the tab pane
+                e.target.classList.contains('ant-tabs-tabpane') &&
+                window.scrollY < TOP_OF_PAGE_RANGE
+              ) {
+                // prevent window from jumping down when tabbing
+                // if already at the top of the page
+                // to help with accessibility when using keyboard navigation
```
Comment (rusackas) [inline superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx:245]: OMG comments! ❤️

File: superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx [hunk 4]
```diff
@@ -350,3 +353,127 @@ test('Should not show the "Edit chart" without proper permissions', () => {
...
+  );
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test('should continue with system tab navigation if dropdown is closed and tab key is pressed', () => {
+    const event = { key: 'Tab', preventDefault: jest.fn() };
+    handleDropdownNavigation(
+      // @ts-ignore
```
Comment (rusackas) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx:376]:
```suggestion
      // @ts-ignore
```
Not blocking, but... how bad is it? 😰

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 5]
```diff
@@ -305,6 +399,7 @@ const SliceHeaderControls = (props: SliceHeaderControlsPropsWithRouter) => {
       case MENU_KEYS.EXPLORE_CHART:
         // eslint-disable-next-line no-unused-expressions
         props.logExploreChart?.(props.slice.slice_id);
+        window.open(props.exploreUrl, '_blank');
```
Comment (rusackas) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:402]: Is popping open in  `_blank` the right default here, or should it open in the same window?

**Comment by rusackas** [comment]
This changes the default behavior of the Edit Chart link, from opening in the same window to a new one. This might open an old can of worms — is it intentional?

File: superset-frontend/src/dashboard/styles.ts [hunk 6]
```diff
@@ -97,3 +111,43 @@ export const chartContextMenuStyles = (theme: SupersetTheme) => css`
...
+    }
+    &:not(.superset-button, .ant-menu-item
+        a, .fave-unfave-icon, .ant-tabs-tabpane, .header-controls span) {
+      &:focus-visible {
+        padding: ${theme.gridUnit / 2}px;
+      }
+    }
+  }
+
+  @keyframes fadeOut {
```
Comment (rusackas) [inline superset-frontend/src/dashboard/styles.ts:136]: Oooooh, fancy! Just another "note to self" that we can add shadows to the theme as well, so they can be appropriate "glows" in dark mode, and be standardized across the UI.

File: superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts (hunk 1 above)
Comment (eschutho) [reply superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts:255]: Yeah this was really tricky because the browser scrolls automatically on focus, and then the field goes under the header and the right click doesn't work. I adjusted for this behavior for real users in the onFocus below but for automations, the wait was the only way to get the browser to catch up. But, that said, there are a few waits earlier in this script that I think we can change to timeouts/assert.is.visible, etc.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx (hunk 4 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx:376]: Oh, it didn't like the type of one of the arguments. I think it was the event that I mocked. I could try to redo it if you think it would be better.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 5 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:402]: I took from the same behavior as the click, which is a new window. But we could change it if we need to?

File: superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts (hunk 1 above)
Comment (eschutho) [reply superset-frontend/cypress-base/cypress/e2e/dashboard/drilltodetail.test.ts:256]: let's see if the above changes make this a net neutral/positive.

**Comment by eschutho** [comment]
> This changes the default behavior of the Edit Chart link, from opening in the same window to a new one. This might open an old can of worms — is it intentional?

No, not intentional. Thanks for flagging that.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx (hunk 4 above)
Comment (rusackas) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/SliceHeaderControls.test.tsx]: Well thanks for fixing... I'm not worried about ts-ignore in tests, so much as the actual application. It's another pile I want to track as tech debt when that dashboard is out in public hands :)

File: superset-frontend/src/dashboard/styles.ts (hunk 6 above)
Comment (rusackas) [reply superset-frontend/src/dashboard/styles.ts]: If you wanna have some fun, you could still add a base box-shadow style to the component with no visible shadow, and add a `transition` property so that the shadow fades on and/or moves when the `focus-visible` pseudoclasss is applied.

Comment (eschutho) [reply superset-frontend/src/dashboard/styles.ts]: I think we're ok with the outline when someone tabs into an element, but it just didn't feel right when clicking on an element, especially when the click bubbles up to an outer element and you have no idea why the entire tab is highlighting.

File: superset-frontend/cypress-base/cypress/e2e/dashboard/editmode.test.ts [hunk 7]
```diff
@@ -483,51 +483,6 @@ describe('Dashboard edit', () => {
...
-      // open main tab and nested tab
-      openTab(0, 0);
-      openTab(1, 1);
-
-      // label Anthony
-      cy.get(
-        '[data-test-chart-name="Top 10 California Names Timeseries"] .line .nv-legend-symbol',
-      )
-        .first()
-        .should('have.css', 'fill', 'rgb(244, 176, 42)');
```
Comment (eschutho) [inline superset-frontend/cypress-base/cypress/e2e/dashboard/editmode.test.ts:530]: I'm just testing what's actually breaking in this file. Will revert

File: superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx [hunk 8]
```diff
@@ -123,7 +123,7 @@ export default function ErrorAlert({
         {!isExpandable && !description && (
           <span
             role="button"
-            tabIndex={0}
+            tabIndex={-1}
```
Comment (kgabryje) [inline superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx:126]: Why are we setting tab index to -1 (here and other places)? Doesn't it have a negative effect on accessibility, since you can't navigate to it with tab anymore?

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 9]
```diff
@@ -231,7 +310,7 @@ const ViewResultsModalTrigger = ({
               <Button
                 buttonStyle="primary"
                 buttonSize="small"
-                onClick={closeModal}
+                onClick={() => setShowModal(false)}
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:313]: is that change necessary? I think it's the same as `closeModal`

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 10]
```diff
@@ -261,10 +340,23 @@ const ViewResultsModalTrigger = ({
...
+  const [drillModalIsOpen, setDrillModalIsOpen] = useState(false);
+  const [selectedKeys, setSelectedKeys] = useState<string[]>([]);
   const [openScopingModal, scopingModal] = useCrossFiltersScopingModal(
     props.slice.slice_id,
   );
 
+  const queryMenuRef: RefObject<any> = useRef(null);
+  const menuRef: RefObject<any> = useRef(null);
+
+  const toggleDropdown = ({ close }: { close?: boolean } = {}) => {
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:354]: Why an object in args? Won't `(close: boolean) => {...}` work the same?

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 11]
```diff
@@ -309,6 +403,7 @@ const SliceHeaderControls = (props: SliceHeaderControlsPropsWithRouter) => {
       case MENU_KEYS.EXPLORE_CHART:
         // eslint-disable-next-line no-unused-expressions
         props.logExploreChart?.(props.slice.slice_id);
+        window.open(props.exploreUrl);
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:406]: Why do we remove a link and use `window.open` instead? The previous approach had all the benefits of using a normal link, like right click context menu

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 12]
```diff
@@ -170,25 +182,91 @@ const dropdownIconsStyles = css`
...
 `;
 
+export const handleDropdownNavigation = (
+  e: React.KeyboardEvent<HTMLElement>,
+  dropdownIsOpen: boolean,
+  menu: React.ReactElement,
+  toggleDropdown: () => void,
+  setSelectedKeys: (keys: string[]) => void,
+) => {
+  if (e.key === NAV_KEYS.tab && !dropdownIsOpen) {
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:192]: Should it be `||`?

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 13]
```diff
@@ -573,15 +691,39 @@ const SliceHeaderControls = (props: SliceHeaderControlsPropsWithRouter) => {
...
+          css={() => css`
             display: flex;
             align-items: center;
           `}
           id={`slice_${slice.slice_id}-controls`}
           role="button"
           aria-label="More Options"
+          tabIndex={0}
+          onClick={() => {
+            toggleDropdown();
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:706]: I think `toggleDropdown` should go to `onVisibleChange` callback of `NoAnimationDropdown`

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 14]
```diff
@@ -573,15 +691,39 @@ const SliceHeaderControls = (props: SliceHeaderControlsPropsWithRouter) => {
...
+            // close unless the dropdown menu is clicked
+            const relatedTarget = e.relatedTarget as HTMLElement;
+            if (
+              dropdownIsOpen &&
+              menuRef?.current?.props.id !== relatedTarget?.id
+            ) {
+              toggleDropdown({ close: true });
+            }
+          }}
+          onKeyDown={e =>
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:718]: I'm wondering if it wouldn't be better to bake all that logic in `NoAnimationDropdown`. We could add those new props (onKeyDown, onBlur) to `NoAnimationDropdown` child by calling `cloneElement`

File: superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx (hunk 8 above)
Comment (eschutho) [reply superset-frontend/src/components/ErrorMessage/ErrorAlert.tsx:126]: Even when navigating to the element, there were no onKeyDown handlers, so you wouldn't be able to do anything after you tabbed to the element, so I removed the ability to tab, because it was out of scope, but I just added the error modal tab functionality since I was here. :)

https://github.com/apache/superset/assets/5186919/6141f4da-c474-47ba-a64e-9bc652b072ec

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 12 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:192]: Here's the expected behavior: if you are tabbing through the page and you are on the slice controls and hit tab when the dropdown is not open, you should jump to the next element in the page instead of interacting with the dropdown elements. So, I think `&&` is right?

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 9 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:313]: thanks for catching! Yes, I removed the `closeModal` method in one of the iterations and forgot to put it back.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 10 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:354]: Yes, you're right, the functionality is the same. It's so that when you call `toggleDropdown({ close: true });` it's more explicit than `toggleDropdown(true);`

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 11 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:406]: When tabbing, the user will land on the Menu item (the li) and then would need to tab again to hit the link in order to open it. This way the action for enter is on the Menu item, so only one tab is needed. It makes sense to have the other context menu benefits, etc, but it's actually really difficult to right click on that element because it gets picked up by the context menu in the chart below. If you have any suggestions about how to add it back though, lmk.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 13 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx]: nice, thanks for that tip! That worked great.

File: superset-frontend/src/features/home/Menu.tsx [hunk 15]
```diff
@@ -306,11 +306,15 @@ export function Menu({
...
+              <GenericLink
+                className="navbar-brand"
+                to={brand.path}
+                tabIndex={-1}
+              >
                 <img src={brand.icon} alt={brand.alt} />
               </GenericLink>
             ) : (
-              <a className="navbar-brand" href={brand.path}>
+              <a className="navbar-brand" href={brand.path} tabIndex={-1}>
```
Comment (eschutho) [inline superset-frontend/src/features/home/Menu.tsx:317]: I'm removing these tabs because they were out of scope. They currently will work and open the homepage when someone tabs and hits enter, but the logo gets larger and distorted when focused and doesn't highlight very well, and the functionality (sending someone to the homepage) for someone who will be using keyboard navigation on this page is not really helpful. We can add it in later, or in this PR if someone feels it's an important step.

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 11 above)
Comment (kgabryje) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:406]: In that case, maybe we could set tabIndex=-1 on that particular menu item, so that tab would focus on the link?

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx [hunk 16]
```diff
@@ -170,25 +182,91 @@ const dropdownIconsStyles = css`
   }
 `;
 
+export const handleDropdownNavigation = (
```
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:185]: I know that this PR is focused on dashboard interactions, but maybe (future iterations) we should add this functionality to the main Dropdown component, so that the keyboard navigation works in all dropdowns?

**Comment by kgabryje** [comment]
When I'm navigating the dropdown with arrows, the whole page scrolls with every click 🙁 And it looks like I can't navigate to Submenus


https://github.com/apache/superset/assets/15073128/9ec868f4-46c5-40bb-acdd-dab953baaeab

File: superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx (hunk 16 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/components/SliceHeaderControls/index.tsx:185]: Yeah, good idea!

**Comment by eschutho** [comment]
> When I'm navigating the dropdown with arrows, the whole page scrolls with every click 🙁 And it looks like I can't navigate to Submenus
> 
Yeah, I didn't add the submenus yet. That's for a separate PR.

File: superset-frontend/src/dashboard/types.ts [hunk 17]
```diff
@@ -237,3 +237,32 @@ export type Slice = {
...
+  SaveModal = 'save_modal',
+  RefreshDashboard = 'refresh_dashboard',
+  AutorefreshModal = 'autorefresh_modal',
+  SetFilterMapping = 'set_filter_mapping',
+  EditProperties = 'edit_properties',
+  EditCss = 'edit_css',
+  ToggleFullscreen = 'toggle_fullscreen',
+  ManageEmbedded = 'manage_embedded',
+  ManageEmailReports = 'manage_email_reports',
+}
```
Comment (eschutho) [inline superset-frontend/src/dashboard/types.ts:268]: Is this file just for typescript types or .. more?

Comment (geido) [reply superset-frontend/src/dashboard/types.ts:268]: What do you mean for "more"?

**Review by geido — APPROVED** [review]

File: superset-frontend/src/dashboard/types.ts (hunk 17 above)
Comment (eschutho) [reply superset-frontend/src/dashboard/types.ts:270]: I thought maybe this file was only to be used for typescript types rather than runtime enums.

File: superset-frontend/cypress-base/cypress/e2e/dashboard/editmode.test.ts [hunk 18]
```diff
@@ -771,7 +771,7 @@ describe('Dashboard edit', () => {
       visitEdit();
     });
 
-    it('should save', () => {
+    it.skip('should save', () => {
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/e2e/dashboard/editmode.test.ts:774]: @eschutho I'm not sure the fate of these tests... should we scrap them, or try to rekindle them?
